### DISCUSSION TITLE
How to get user information in middleware plugin?

### DISCUSSION THREAD

**Discussion by sivanzheng** [verdaccio/verdaccio#4747] [Q&A] [answered]
I want to implement a middleware plugin that can log package changes, but I don't know how to get user information in it.

**Comment by sivanzheng** [comment]
I try to get user information from req.remote_user but i got undefined
```ts
    public register_middlewares(
        app: Application,
        auth: IBasicAuth<CustomConfig>,
        _storage: IStorageManager<CustomConfig>
    ): void {
        const router = Router();

        // Log request details to console and file
        const logRequest = (req: any) => { console.log(req.remote_user) }
            
        router.put('/:package', (req: Request, res: Response, next: NextFunction) => {
            logRequest(req);
            next();
        });

        app.use('/', router);
    }
```

**Reply by juanpicado** [reply]
> I try to get user information from req.remote_user but i got undefined

After checking the code register the middleware happens before the token is actually set to `req.remote_user`, so probably you need to do to it by your own.

<img width="627" alt="Screenshot 2024-08-01 at 18 38 28" src="https://github.com/user-attachments/assets/18c32ec1-4bac-4fba-8b93-ccf3cc83c73f">

You may use `route.use(auth.apiJWTmiddleware());` , that would set `req.remote_user` at the request level, I am not 100% sure if may work but give it a try.

**Reply by sivanzheng** [reply, marked as answer]
Thank you very much. After I checked the source code of Verdaccio, I also tried to use the `apiJWTmiddleware` method. After your reminder, I succeeded. This is a useful method.
```ts
router.put('/:package', (req, res, next) => {
    if (auth.apiJWTmiddleware) {
        auth.apiJWTmiddleware()(req, res, () => {
            const remote_user = (req as any).remote_user as RemoteUser;
            console.log('remote', remote_user)
            logRequest(req);
            next();
        });
    } else {
        logRequest(req);
        next();
    }
});
```
